File: rust/src/model/metrics/action.rs
use crate::model::domain::{Cost, Period, Purchase, Sale};

#[derive(Debug, Clone)]
pub enum Action {
    ProcessPeriod,
    ProcessYear,
    Purchase(Purchase),
    Sale(Sale),
    Cost(Cost),
    ClosePeriod(Period),
    OpenPeriod(Period),
    PeriodReport(Period),
    YearReport(Period),
}

impl Action {
    pub fn code(&self) -> &'static str {
        match self {
            Action::ProcessPeriod => { "PROCESS_PERIOD" }
            Action::ProcessYear => { "PROCESS_YEAR" }
            Action::Purchase(_) => { "PURCHASE" }
            Action::Sale(_) => { "SALE" }
            Action::Cost(_) => { "COST" }
            Action::ClosePeriod(_) => { "CLOSE_PERIOD" }
            Action::OpenPeriod(_) => { "OPEN_PERIOD" }
            Action::PeriodReport(_) => { "PERIOD_REPORT" }
            Action::YearReport(_) => { "YEAR_REPORT" }
        }
    }
}

File: golang/model/metrics/action.go
package metrics

import "bench-pg-go/model/domain"

type Action interface {
	Type() string
}

type ProcessYear struct{}

func NewProcessYear() ProcessYear {
	return ProcessYear{}
}

func (ProcessYear) Type() string {
	return "PROCESS_YEAR"
}

type ProcessPeriod struct{}

func NewProcessPeriod() ProcessPeriod {
	return ProcessPeriod{}
}

func (ProcessPeriod) Type() string {
	return "PROCESS_PERIOD"
}

type Operation struct {
	Operation domain.Operation
}

func NewOperationAction(operation domain.Operation) Operation {
	return Operation{operation}
}

func (o Operation) Type() string {
	return o.Operation.Code()
}
